packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol

```diff
@@ -145,45 +142,46 @@ contract DirectGovernanceFarmer is ReentrancyGuard {
         address token,
         uint256 amount,
         uint256 durationWeeks
-    ) external nonReentrant returns (uint256 stakingShareId) {
+    ) internal returns (uint256 stakingShareId) {
+        DirectGovernanceData storage data = directGovernanceStorage();
         // DAI / USDC / USDT / Ubiquity Dollar
         require(
             isMetaPoolCoin(token),
             "Invalid token: must be DAI, USD Coin, Tether, or Ubiquity Dollar"
         );
-        require(amount > 0, "amount must be positive vale");
+        require(amount > 0, "amount must be positive value");
         require(
             durationWeeks >= 1 && durationWeeks <= 208,
             "duration weeks must be between 1 and 208"
         );
 
         //Note, due to USDT implementation, normal transferFrom does not work and have an error of "function returned an unexpected amount of data"
-        //require(IERC20(token).transferFrom(msg.sender, address(this), amount), "sender cannot transfer specified fund");
         IERC20(token).safeTransferFrom(msg.sender, address(this), amount);
 
-        address staking = manager.stakingContractAddress();
-        address stakingShare = manager.stakingShareAddress();
+        address staking = address(data.manager.stakingContractAddress());
+        address stakingShare = address(data.manager.stakingShareAddress());
 
-        uint256 lpAmount; //UAD3CRVf
         //[Ubiquity Dollar, DAI, USDC, USDT]
         uint256[4] memory tokenAmounts = [
-            token == ubiquityDollar ? amount : 0,
-            token == token0 ? amount : 0,
-            token == token1 ? amount : 0,
-            token == token2 ? amount : 0
+            token == address(data.ubiquityDollar) ? amount : 0,
+            token == data.token0 ? amount : 0,
+            token == data.token1 ? amount : 0,
+            token == data.token2 ? amount : 0
         ];
 
         //STEP1: add DAI, USDC, USDT or Ubiquity Dollar into metapool liquidity and get UAD3CRVf
-        IERC20(token).safeIncreaseAllowance(depositZapUbiquityDollar, amount);
-        lpAmount = IDepositZap(depositZapUbiquityDollar).add_liquidity(
-            ubiquity3PoolLP,
-            tokenAmounts,
-            0
+        IERC20(token).safeIncreaseAllowance(
+            data.depositZapUbiquityDollar,
+            amount
         );
+        uint256 lpAmount = IDepositZap(data.depositZapUbiquityDollar)
+            .add_liquidity(data.ubiquity3PoolLP, tokenAmounts, 0);
 
         //STEP2: stake UAD3CRVf to Staking
+
         //TODO approve token to be transferred to Staking contract
```

Comment:
since approve was implemented TODO comment can be deleted

Reply:
no, no deleting, this will guideful on a future PR, but not on this, we need this to get in, then we check it out this one

Reply:
some comments are misleading from the older commits but im recycling on a new PR

Reply:
all right